contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol

```diff
@@ -0,0 +1,335 @@
+            require(t2eSwapPaths[token].length != 0, "no path is exists for t2e");
+        }
+
+        token.safeApprove(address(uniswapRouter), 2**255);
+
+        setDecimals(token);
+
+        emit TokenListed(token, true);
+    }
+
+    function delistToken(IERC20 token) external {
+        onlyOperator();
+        require(tokenListed[token], "token is not listed");
+        delete tokenListed[token];
+        // clear all paths data
+        delete t2eSwapPaths[token];
+        delete e2tSwapPaths[token];
+
+        token.safeApprove(address(uniswapRouter), 0);
+        emit TokenListed(token, false);
+    }
+
+    function removePath(
+        IERC20 token,
+        bool isEthTotoken,
+        uint256 index
+    ) external {
+        onlyOperator();
+        address[][] storage allPaths;
+        if (isEthTotoken) {
+            allPaths = e2tSwapPaths[token];
+        } else {
+            allPaths = t2eSwapPaths[token];
+        }
+        require(index < allPaths.length, "invalid index");
+        address[] memory path = allPaths[index];
+        allPaths[index] = allPaths[allPaths.length - 1];
+        allPaths.pop();
+
+        emit TokenPathAdded(token, path, isEthTotoken, false);
+    }
+
+    function enableTrade() external returns (bool) {
+        onlyAdmin();
+        tradeEnabled = true;
+        emit TradeEnabled(true);
+        return true;
+    }
+
+    function disableTrade() external returns (bool) {
+        onlyAlerter();
+        tradeEnabled = false;
+        emit TradeEnabled(false);
+        return true;
+    }
+
+    /**
+     *   @dev called by kybernetwork to get settlement rate
+     */
+    function getConversionRate(
+        IERC20 src,
+        IERC20 dest,
+        uint256 srcQty,
+        uint256 /* blockNumber */
+    ) external override view returns (uint256) {
+        if (!isValidTokens(src, dest)) return 0;
+        if (!tradeEnabled) return 0;
+        if (srcQty == 0) return 0;
+
+        (uint256 rate, ) = calcUniswapConversion(src, dest, srcQty);
+        return rate;
+    }
+
+    function addPath(
+        IERC20 token,
+        address[] memory path,
+        bool isEthToToken
+    ) public {
+        onlyOperator();
+        address[][] storage allPaths;
+
+        require(path.length >= 2, "path is too short");
+        if (isEthToToken) {
+            require(path[0] == weth, "start address of path is not weth");
+            require(
+                path[path.length - 1] == address(token),
+                "end address of path is not token"
+            );
+            allPaths = e2tSwapPaths[token];
+        } else {
+            require(path[0] == address(token), "start address of path is not token");
+            require(path[path.length - 1] == weth, "end address of path is not weth");
+            allPaths = t2eSwapPaths[token];
+        }
+        // verify the pair is existed and the pair has liquidity
+        for (uint256 i = 0; i < path.length - 1; i++) {
```

Comment:
Could we use a simpler way to verify if path has liquidity by getting rate with 1 token?

Reply:
1 twei or one token?

had the same question
it 1 twei and 2nd token is worth lest amount out would be 0.